File: cpp/dsa/data-structures/stack-queue/sliding-window-maximum.cpp
vector<int> maxSlidingWindow(vector<int>& nums, int k) {
    deque<int> dq;
    vector<int> res;
    int left = 0;

    for (int right = 0; right < nums.size(); right++) {
        // 1. Expand
        while (!dq.empty() && nums[dq.back()] <= nums[right]) {
            dq.pop_back();
        }
        dq.push_back(right);

        // 2. Condition
        if (right - left + 1 == k) {
            // 3. Record
            res.push_back(nums[dq.front()]);

            // 4. Slide
            if (dq.front() == left) {
                dq.pop_front();
            }
            left++;
        }
    }
    return res;
}

File: go/dsa/data-structures/stack-queue/sliding-window-maximum.go
func maxSlidingWindow(nums []int, k int) []int {
    dq := []int{}
    res := []int{}
    left := 0

    for right := 0; right < len(nums); right++ {
        // 1. Expand
        for len(dq) > 0 && nums[dq[len(dq)-1]] <= nums[right] {
            dq = dq[:len(dq)-1]
        }
        dq = append(dq, right)

        // 2. Condition
        if right - left + 1 == k {
            // 3. Record
            res = append(res, nums[dq[0]])

            // 4. Slide
            if dq[0] == left {
                dq = dq[1:]
            }
            left++
        }
    }
    return res
}